## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [3]:
# Import dependencies
import pandas as pd
from pathlib import Path

# Create a reference to the CSV file
csv_path = Path("Resources/client_dataset.csv")

# Read the CSV file in as a Pandas DataFrame
orders_df = pd.read_csv(csv_path, low_memory=False)

# Print the first five rows of the DataFrame's data to the screen
orders_df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [5]:
# View the column names in the data
orders_df.columns


Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [6]:
# Use the describe function to gather some basic statistics
orders_df.describe()


client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [11]:
# EXPLORATION
# Use this space to do any additional research
# and familiarize yourself with the data.

# Get the number of unique client_id values in the data
orders_df['client_id'].nunique()

999

In [49]:
# EXPLORATION

# Group the data by the 'client_id' column to find the number of orders per client
clients_df_pivot = orders_df.pivot_table(index='client_id', values='order_id', aggfunc='count')
clients_df_pivot.head()

order_id
client_id          
10033            19
10091            14
10218            28
10351            50
10435           143

In [50]:
# EXPLORATION

# Group the data by the 'order_id' column to find the number of clients per order
orders_df_pivot = orders_df.pivot_table(index='order_id', values='client_id', aggfunc='count')
orders_df_pivot.head()

client_id
order_id           
1000886           2
1001431           9
1003003           4
1003666          10
1004537          10

In [51]:
# EXPLORATION

# Here we confirm the an order_id is unique to a client_id but shows up multiple times in the dataset since it is associated with individual line items/products
orders_products_df_pivot = orders_df.pivot_table(index='order_id', values='item_id', aggfunc='count')
orders_products_df_pivot.head()

item_id
order_id         
1000886         2
1001431         9
1003003         4
1003666        10
1004537        10

In [12]:
# EXPLORATION

# Group the data by the 'client_id' column to find the number of orders per client, using just value counts
orders_df['client_id'].value_counts()

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
        ... 
70734      4
32001      4
77583      3
76017      3
94987      1
Name: count, Length: 999, dtype: int64

In [10]:
# EXPLORATION

# Find the number of unique order_id values in the dataset
orders_df['order_id'].nunique()

10000

In [15]:
# EXPLORATION

# Find the maximum number of times a single order_id appears in the dataset
orders_df['order_id'].value_counts().max()

10

In [17]:
# What three item categories had the most entries?
orders_df['category'].value_counts().head(3)


category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [20]:
# For the category with the most entries, which subcategory had the most entries?
orders_df[orders_df['category'] == 'consumables']['subcategory'].value_counts().head(1)

subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [21]:
# Which five clients had the most entries in the data?
orders_df['client_id'].value_counts().head(5)


client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [24]:
# Store the client ids of those top 5 clients in a list.
top_5_clients_list = orders_df['client_id'].value_counts().head(5).index.tolist()
top_5_clients_list


[33615, 66037, 46820, 38378, 24741]

In [25]:
# How many total units (the qty column) did the client with the most entries order order?
orders_df[orders_df['client_id'] == top_5_clients_list[0]]['qty'].sum()


64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [27]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
orders_df['subtotal'] = orders_df['unit_price'] * orders_df['qty']
orders_df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number   subtotal  
0     1096.80     762.71         7.50  105            1  115164.00  
1       24.95      15.09         1.49   21            0     523.95  
2       13.52       7.86         1.68   39            6     527.28  
3       36.42      24.85         1.23   29            3    1056.18  
4      195.10     108.17        46.43   20            1    3902.00

In [30]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
orders_df['shipping'] = 0 # Initialize the column with zeros
orders_df.loc[orders_df['unit_weight'] > 50, 'shipping'] = orders_df['unit_weight'] * 7
orders_df.loc[orders_df['unit_weight'] <= 50, 'shipping'] = orders_df['unit_weight'] * 10
orders_df.head()



/var/folders/rr/f4rnf7xj5tj1x2gknbjs4mxm0000gn/T/ipykernel_92708/1450780444.py:5: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value '[ 75.   14.9  16.8 ...  41.6 180.4  20.7]' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.
  orders_df.loc[orders_df['unit_weight'] <= 50, 'shipping'] = orders_df['unit_weight'] * 10


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number   subtotal  shipping  
0     1096.80     762.71         7.50  105            1  115164.00      75.0  
1       24.95      15.09         1.49   21            0     523.95      14.9  
2       13.52       7.86         1.68   39            6     527.28      16.8  
3       36.42      24.85         1.23   29            3    1056.18      12.3  
4      195.10     108.17        46.43   20            1    3902.00     464.3

In [34]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
orders_df['tax'] = 0.0925
orders_df['total'] = orders_df['subtotal'] + orders_df['shipping']
orders_df['total'] = orders_df['total'] + (orders_df['total'] * orders_df['tax'])
orders_df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...         subcategory unit_price unit_cost  \
0          17        2023  ...            wall art    1096.80    762.71   
1          20        2023  ...                pens      24.95     15.09   
2           4        2023  ...  project management      13.52      7.86   
3          17        2023  ...                pens      36.42     24.85   
4          21        2023  ...                misc     195.10    108.17   

   unit_weight  qty  line_number   subtotal  shipping          total     tax  
0         7.50  105            1  115164.00      75.0  125898.607500  0.0925  
1         1.49   21            0     523.95      14.9     588.693625  0.0925  
2         1.68   39            6     527.28      16.8     594.407400  0.0925  
3         1.23   29            3    1056.18      12.3    1167.314400  0.0925  
4        46.43   20            1    3902.00     464.3    4770.182750  0.0925  

[5 rows x 22 columns]

In [35]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
orders_df['cost'] = orders_df['unit_cost'] * orders_df['qty'] + orders_df['shipping']
orders_df.head()



first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   
3          17        2023  ...      36.42     24.85        1.23   29   
4          21        2023  ...     195.10    108.17       46.43   20   

   line_number   subtotal  shipping          total     tax      cost  
0            1  115164.00      75.0  125898.607500  0.0925  80159.55  
1            0     523.95      14.9     588.693625  0.0925    331.79  
2            6     527.28      16.8     594.407400  0.0925    323.34  
3            3    1056.18      12.3    1167.314400  0.0925    732.95  
4            1    3902.00     464.3    4770.182750  0.0925   2627.70  

[5 rows x 23 columns]

In [37]:
# Create a column for the profit of each line using line cost and line price
orders_df['profit'] = orders_df['total'] - orders_df['cost']
orders_df.head()



first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   
3          17        2023  ...     24.85        1.23   29            3   
4          21        2023  ...    108.17       46.43   20            1   

    subtotal  shipping          total     tax      cost        profit  
0  115164.00      75.0  125898.607500  0.0925  80159.55  45739.057500  
1     523.95      14.9     588.693625  0.0925    331.79    256.903625  
2     527.28      16.8     594.407400  0.0925    323.34    271.067400  
3    1056.18      12.3    1167.314400  0.0925    732.95    434.364400  
4    3902.00     464.3    4770.182750  0.0925   2627.70   2142.482750  

[5 rows x 24 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [45]:
orders_df.groupby('order_id').sum()['total'].head()

order_id
1000886     97297.143225
1001431    170268.288150
1003003      9884.404675
1003666    721014.777800
1004537     88184.928475
Name: total, dtype: float64

In [43]:
# Check your work using the totals above
# Order ID 2742071 had a total price of \$152,811.89
# Order ID 2173913 had a total price of \$162,388.71
# Order ID 6128929 had a total price of \$923,441.25

print(orders_df[orders_df['order_id'] == 2742071]['total'].sum())
print(orders_df[orders_df['order_id'] == 2173913]['total'].sum())
print(orders_df[orders_df['order_id'] == 6128929]['total'].sum())

139844.51202499997
119838.35704999999
752748.745475


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [44]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
for client in top_5_clients_list:
    print(client, orders_df[orders_df['client_id'] == client]['total'].sum())



33615 6392918.349675
66037 8746255.3054
46820 8006332.531775
38378 9172866.143875003
24741 76678479.053625


In [ ]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 


In [ ]:
# Format the data and rename the columns to names suitable for presentation.

# Define the money columns. 


# Define a function that converts a dollar amount to millions.


# Apply the currency_format_millions function to only the money columns. 


# Rename the columns to reflect the change in the money format. 




In [ ]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
